File: Labs/Lab8/Main.java
import java.util.Arrays;

public class Main {
    public int coinChange(int[] coins, int amount) {
        int[] arr = new int[amount + 1];
        Arrays.fill(arr, Integer.MAX_VALUE);
        arr[0] = 0;

        for (int i = 0; i <= amount; i++) {
            for (int coin: coins) {
                if (i - coin >= 0) {
                    if (arr[i - coin] != Integer.MAX_VALUE
                            && 1 + arr[i - coin] < arr[i]) {
                        arr[i] = 1 + arr[i - coin];
                    }
                }
            }
        }

        if (arr[amount] == Integer.MAX_VALUE){
            return -1;
        }
        return arr[amount];
    }
    public static void main(String[] args) {
        int[] coins = new int[]{186, 419, 83, 408};
        int amount = 6249;
        Main obj = new Main();
        System.out.println(obj.coinChange(coins, amount));
    }
}

File: Labs/Lab8/main.py
import sys
from typing import List


def coinChange(self, coins: List[int], amount: int) -> int:
    arr = [sys.maxsize] * (amount + 1)
    arr[0] = 0
    for i in range(amount + 1):
        for coin in coins:
            if i - coin >= 0:
                if 1 + arr[i - coin] < arr[i]:
                    arr[i] = 1 + arr[i-coin]
    if arr[amount] == sys.maxsize:
        return -1
    return arr[amount]


coins = [186,419,83,408]
amount = 6249
coin_amount = coinChange(coinChange, coins, amount)
print(coin_amount)
